src/libraries/System.Net.Sockets/tests/FunctionalTests/Connect.cs

```diff
@@ -204,4 +204,133 @@ public sealed class ConnectEap : Connect<SocketHelperEap>
     }
+
+    public sealed class ConnectCancellableTask : Connect<SocketHelperCancellableTask>
+    {
+        public ConnectCancellableTask(ITestOutputHelper output) : base(output) { }
+
+        [Fact]
+        public async Task ConnectEndPoint_Precanceled_Throws()
+        {
+            EndPoint ep = new IPEndPoint(IPAddress.Parse("1.2.3.4"), 1);
+
+            using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                var cts = new CancellationTokenSource();
+                cts.Cancel();
+
+                await Assert.ThrowsAnyAsync<OperationCanceledException>(async () => await client.ConnectAsync(ep, cts.Token));
+            }
+        }
+
+        [Fact]
+        public async Task ConnectAddressAndPort_Precanceled_Throws()
+        {
+            using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                var cts = new CancellationTokenSource();
+                cts.Cancel();
+
+                await Assert.ThrowsAnyAsync<OperationCanceledException>(async () => await client.ConnectAsync(IPAddress.Parse("1.2.3.4"), 1, cts.Token));
+            }
+        }
+
+        [Fact]
+        public async Task ConnectMultiAddressAndPort_Precanceled_Throws()
+        {
+            using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                var cts = new CancellationTokenSource();
+                cts.Cancel();
+
+                await Assert.ThrowsAnyAsync<OperationCanceledException>(async () => await client.ConnectAsync(new IPAddress[] { IPAddress.Parse("1.2.3.4"), IPAddress.Parse("1.2.3.5") }, 1, cts.Token));
+            }
+        }
+
+        [Fact]
+        public async Task ConnectHostNameAndPort_Precanceled_Throws()
+        {
+            using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                var cts = new CancellationTokenSource();
+                cts.Cancel();
+
+                await Assert.ThrowsAnyAsync<OperationCanceledException>(async () => await client.ConnectAsync("1.2.3.4", 1, cts.Token));
+            }
+        }
+
+        [Fact]
+        [OuterLoop("Uses Task.Delay")]
+        [PlatformSpecific(TestPlatforms.Windows)]   // Linux will not even attempt to connect to the invalid IP address
+        public async Task ConnectEndPoint_CancelDuringConnect_Throws()
+        {
+            EndPoint ep = new IPEndPoint(IPAddress.Parse("1.2.3.4"), 1);
+
+            using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                var cts = new CancellationTokenSource();
+
+                ValueTask t = client.ConnectAsync(ep, cts.Token);
+
+                // Delay cancellation a bit to try to ensure the OS actually attempts to connect
+                cts.CancelAfter(100);
```

Comment:
Are there any tests on Linux where we actually cancel the connect?  I only skimmed, but it seems like all the ones that cancel the connect are Windows-specific.

Reply:
I've modified these tests to actually work on linux, by making them work the way they existing CancelConnectAsync tests work.